crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//! and the number of vaccinations is the value. But, if we do that in the most
+//! obvious way, we end up with just the number of vaccinations as the result,
+//! whereas we probably want to know the year and month that that number of
+//! occurred as well.
+//!
+//! One way to recover the year and month is to join against the original data.
+//! We can do it without another join by defining a type that is ordered by
+//! vaccination count but also contains the year and month.  Taking advantage of
+//! how Rust derives [`Ord`] lexicographically, that's as simple as:
+//!
+//! ```rust
+//! # use size_of::SizeOf;
+//!
+//! #[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, SizeOf)]
+//! struct VaxMonthly {
+//!     count: u64,
+//!     year: i32,
+//!     month: u8,
+//! }
+//! ```
+//!
+//! We can transform our monthly totals from a `(country, year, month)`
+//! key and `vaccinations` value in `country` value and `VaxMonthly` value
+//! with:
+//!
+//! ```ignore
+//!     monthly_totals
+//!         .map_index(|((c, y, m), sum)| {
+//!             (
+//!                 c.clone(),
+//!                 VaxMonthly {
+//!                     count: *sum as u64,
+//!                     year: *y,
+//!                     month: *m,
+//!                 },
+//!             )
+//!         })
+//! ```
+//!
+//! Getting the top three months for each country and printing them is
+//! then as simple as:
+//!
+//! ```ignore
+//!         .topk_desc(3)
+//!         .inspect(|data| {
+//!             data.iter()
+//!                 .for_each(|(c, VaxMonthly { count, year, month }, w)| {
+//!                     println!("{c:16} {year}-{month:02} {count:10}: {w:+}")
+//!                 })
+//!         });
+//! ```
+//!
+//! The complete program is in `tutorial8.rs`.  When we run it, it outputs:
+//!
+//! ```text
+//! England          2021-03   11861175: +1
+//! England          2021-05   13772946: +1
+//! England          2021-12   14801300: +1
+//! Northern Ireland 2021-05     394047: +1
+//! Northern Ireland 2021-04     436870: +1
+//! Northern Ireland 2021-12     489059: +1
+//! Scotland         2021-06    1244155: +1
+//! Scotland         2021-05    1272194: +1
+//! Scotland         2021-12    1388549: +1
+//! Wales            2021-04     707714: +1
+//! Wales            2021-12     822945: +1
+//! Wales            2021-03     836844: +1
+//! ```
+//!
+//! ### Vaccination rates
+//!
+//! Suppose we want to compare the population in each country with the
+//! number of vaccinations given, that is, to calculate a vaccination
+//! rate.  Our input data contains a `total_vaccinations_per_hundred`
+//! column that reports this information, but let's calculate it ourselves
+//! to demonstrate how it might be done in DBSP.
+//!
+//! We need to know the population in each country.  That's a separate
+//! input table.  Obviously, population changes over time, but for
+//! simplicity let's just pick a single figure for each one.
```

Comment:
Is this last comment necessary? I don't think we require the population to _not_ change anywhere.

Reply:
I was trying to explain why my population table didn't have a date column.

I changed this and the following paragraph to:

```
//! We need to know the population in each country.  Let's add a new input
//! source for population by country.  Since that's naturally a set of key-value
//! pairs, we'll make it an indexed Z-set input source instead of a plain Z-set.
//! The first step is to make our circuit builder construct the source and
//! return its handle:
```